### ISSUE TITLE
Bug with --local_address  option and  "_on_before_issue() " function 

### ISSUE THREAD

**Issue by stevebovy** [Neilpang/acme.sh#1303]
The "port-checking" logic for the --local-address option has a bug

in the "_on_before_issue() {"  function  

The following lines ( approx line-no 3015 )  is not correct  

      if [ -z "$netprc" ]; then
        netprc="$(echo "$_netprc" | grep "$LOCAL_ANY_ADDRESS")"
      fi

They should be changed as follows :

      if [ "$netprc" != "" ]; then
        netprc="$(echo "$_netprc" | grep "$LOCAL_ANY_ADDRESS")"
      fi

Steps to reproduce
------------------

--issue --test --standalone -d home.miscbitbag.org --local-address $(wget -O - -q http://myip.dnsomatic.com/)

Debug log
-----------------

[Fri Feb 23 15:01:35 PST 2018] Standalone mode.
[Fri Feb 23 15:01:35 PST 2018] _checkport='80'
[Fri Feb 23 15:01:35 PST 2018] _checkaddr='75.83.231.181'
[Fri Feb 23 15:01:35 PST 2018] ss exists=1
[Fri Feb 23 15:01:35 PST 2018] netstat exists=0
[Fri Feb 23 15:01:35 PST 2018] Using: netstat
[Fri Feb 23 15:01:35 PST 2018] tcp        0      0 127.0.0.1:80            0.0.0.0:*               LISTEN
tcp        0      0 192.168.50.1:80         0.0.0.0:*               LISTEN
[Fri Feb 23 15:01:35 PST 2018] tcp port 80 is already used by
[Fri Feb 23 15:01:35 PST 2018] Please stop it first
[Fri Feb 23 15:01:35 PST 2018] _on_before_issue.

```
acme.sh  --issue .....   --debug 2
```

**Comment by Neilpang** [comment]
why ?

**Comment by stevebovy** [comment]
>> I am going to add comments to the following code  to explain my reasoning;
>> Maybe my solution is not 100 % correct :)
...
 
     # check for open ports
      _netprc="$(_ss "$_checkport" | grep "$_checkport")"
    # $_checkaddr is  --local_address
    # check for open ports on "--local_address i.e  { --local_address:80 } "
      netprc="$(echo "$_netprc" | grep "$_checkaddr")"

      # when netprc == "" and  $_checkaddr then no more check are needed
      # if [ -z ] does not work correctly  when $_checkaddr ; change if statement to make it work

      if [ "$netprc" != "" ]; then
        _debug "netprc: checking any address"
        netprc="$(echo "$_netprc" | grep "$LOCAL_ANY_ADDRESS")"
      fi

      if [ "$netprc" ]; then
        _err "$netprc"
        _err "tcp port $_checkport is already used by $(echo "$netprc" | cut -d : -f 4)"
        _err "Please stop it first"
        return 1
      fi

...

**Comment by stevebovy** [comment]
Thanks for listening;  now on further analysis I see my solution was incorrect;

It should be as follows:

...
     # $_checkaddr = --local_address ( look for matching address )
      netprc="$(echo "$_netprc" | grep "$_checkaddr")"
     # if no matching address; then no more checks are needed
      if [ -z "$netprc" && "$_checkaddr" == "" ]; then
        _debug "netprc: checking any address"
        netprc="$(echo "$_netprc" | grep "$LOCAL_ANY_ADDRESS")"
      fi
...

**Comment by stevebovy** [comment]
You should not close the issue until you receive my answer !!!

 

# $_checkaddr = --local_address ( look for matching local-address )

  netprc="$(echo "$_netprc" | grep "$_checkaddr")"

# if no matching local-address; then no more checks are needed

  if [ -z "$netprc" ] && [ "$_checkaddr" = "" ]; then

    _debug "netprc: checking any address"

    netprc="$(echo "$_netprc" | grep "$LOCAL_ANY_ADDRESS")"

  fi

**Comment by stevebovy** [comment]
Thanks for listening; now on further analysis I see my solution was incorrect;

It should be as follows:

 # $_checkaddr = --local_address ( look for matching local-address )
  netprc="$(echo "$_netprc" | grep "$_checkaddr")"
 # if no matching local-address; then no more checks are needed
  if [ -z "$netprc" ] && [ "$_checkaddr" = "" ]; then
    _debug "netprc: checking any address"
    netprc="$(echo "$_netprc" | grep "$LOCAL_ANY_ADDRESS")"
  fi

**Comment by Neilpang** [comment]
The `local-address` features was introduced a long time ago.  And I tested it on my router, it works as expected.

paste your `--debug 2` log, if you have any problems.

**Comment by stevebovy** [comment]
Thanks; but you are not listening 

 

Did you ever test it with port 80 open on a different interface ???

 

      Interface-x:port-80 

 

      Local-address-interface:port-80 

    

Your check logic has a design flaw

**Comment by stevebovy** [comment]
I apologize for my poor communication skills  ;  let me try again to explain the situation  :)  

 

If we are using   local-address= “interface-x” ;

 

THEN   “interface-x”    is the   ONLY THING  THAT  MATTERS;   

 

ALL OTHER  interfaces  ARE  IRELEVENT AND SHOULD NOT  BE CHECKED   

 

>> Here are  my results  without   my fix :

 

/mnt/Optware/acmesh/acme.sh --renew --force --test

[Sat Feb 24 10:04:07 PST 2018] Using stage ACME_DIRECTORY: https://acme-staging.api.letsencrypt.org/directory

[Sat Feb 24 10:04:07 PST 2018] Renew: 'home.miscbitbag.org'

[Sat Feb 24 10:04:09 PST 2018] Standalone mode.

 

tcp        0      0 127.0.0.1:80                0.0.0.0:*               LISTEN

tcp        0      0 192.168.50.1:80         0.0.0.0:*               LISTEN

 

[Sat Feb 24 10:04:09 PST 2018] tcp port 80 is already used by

[Sat Feb 24 10:04:10 PST 2018] Please stop it first

[Sat Feb 24 10:04:10 PST 2018] _on_before_issue.

 

>> As you can see above none of the  interface-matches  are a problem because they are not  “interface-x”  

 

NOTE: strange under these circumstances ;  the following code does not seem to have any affect :

 

        >>  By $(echo "$netprc" | cut -d : -f 4)

**Comment by stevebovy** [comment]
>> Here is my debug log 
>> Note I added a few debug statements 

/mnt/Optware/acmesh/acme.sh --renew --force --test  --log --debug 3 --syslog 7 --log-level 2 --local-address 75.83.231.181
[Sat Feb 24 10:36:00 PST 2018] readlink exists=0
[Sat Feb 24 10:36:00 PST 2018] dirname exists=0
[Sat Feb 24 10:36:00 PST 2018] Lets find script dir.
[Sat Feb 24 10:36:00 PST 2018] _SCRIPT_='/mnt/Optware/acmesh/acme.sh'
[Sat Feb 24 10:36:00 PST 2018] _script='/tmp/mnt/Optware/acmesh/acme.sh'
[Sat Feb 24 10:36:00 PST 2018] _script_home='/tmp/mnt/Optware/acmesh'
[Sat Feb 24 10:36:00 PST 2018] Using default home:/root/.acme.sh
[Sat Feb 24 10:36:00 PST 2018] Using config home:/root/.acme.sh
[Sat Feb 24 10:36:00 PST 2018] OK
[Sat Feb 24 10:36:00 PST 2018] 1:LOG_FILE='/root/.acme.sh/acme.sh.log'
[Sat Feb 24 10:36:00 PST 2018] OK
[Sat Feb 24 10:36:00 PST 2018] 2:LOG_LEVEL='2'
[Sat Feb 24 10:36:00 PST 2018] logger exists=0
[Sat Feb 24 10:36:00 PST 2018] OK
[Sat Feb 24 10:36:00 PST 2018] 3:SYS_LOG='7'
[Sat Feb 24 10:36:00 PST 2018] LE_WORKING_DIR='/root/.acme.sh'
https://github.com/Neilpang/acme.sh
v2.7.7
[Sat Feb 24 10:36:00 PST 2018] Using config home:/root/.acme.sh
[Sat Feb 24 10:36:00 PST 2018] Using stage ACME_DIRECTORY: https://acme-staging.api.letsencrypt.org/directory
[Sat Feb 24 10:36:00 PST 2018] ACME_DIRECTORY='https://acme-staging.api.letsencrypt.org/directory'
[Sat Feb 24 10:36:00 PST 2018] _ACME_SERVER_HOST='acme-staging.api.letsencrypt.org'
[Sat Feb 24 10:36:00 PST 2018] CA_CONF='/root/.acme.sh/ca/acme-staging.api.letsencrypt.org/ca.conf'
[Sat Feb 24 10:36:01 PST 2018] DOMAIN_PATH='/root/.acme.sh/home.miscbitbag.org'
[Sat Feb 24 10:36:01 PST 2018] Renew: 'home.miscbitbag.org'
[Sat Feb 24 10:36:01 PST 2018] initpath again.
[Sat Feb 24 10:36:01 PST 2018] Using config home:/root/.acme.sh
[Sat Feb 24 10:36:01 PST 2018] ACME_DIRECTORY='https://acme-staging.api.letsencrypt.org/directory'
[Sat Feb 24 10:36:01 PST 2018] _ACME_SERVER_HOST='acme-staging.api.letsencrypt.org'
[Sat Feb 24 10:36:01 PST 2018] CA_CONF='/root/.acme.sh/ca/acme-staging.api.letsencrypt.org/ca.conf'
[Sat Feb 24 10:36:01 PST 2018] Using ACME_DIRECTORY: https://acme-staging.api.letsencrypt.org/directory
[Sat Feb 24 10:36:01 PST 2018] _init api for server: https://acme-staging.api.letsencrypt.org/directory
[Sat Feb 24 10:36:01 PST 2018] GET
[Sat Feb 24 10:36:01 PST 2018] url='https://acme-staging.api.letsencrypt.org/directory'
[Sat Feb 24 10:36:01 PST 2018] timeout=
[Sat Feb 24 10:36:01 PST 2018] curl exists=0
[Sat Feb 24 10:36:01 PST 2018] mktemp exists=0
[Sat Feb 24 10:36:01 PST 2018] wget exists=0
[Sat Feb 24 10:36:01 PST 2018] _CURL='curl -L --silent --dump-header /root/.acme.sh/http.header  --trace-ascii /tmp/tmp.LMOiNLvRtP  -g '
[Sat Feb 24 10:36:02 PST 2018] ret='0'
[Sat Feb 24 10:36:02 PST 2018] response='{
  "key-change": "https://acme-staging.api.letsencrypt.org/acme/key-change",
  "meta": {
    "terms-of-service": "https://letsencrypt.org/documents/LE-SA-v1.2-November-15-2017.pdf"
  },
  "new-authz": "https://acme-staging.api.letsencrypt.org/acme/new-authz",
  "new-cert": "https://acme-staging.api.letsencrypt.org/acme/new-cert",
  "new-reg": "https://acme-staging.api.letsencrypt.org/acme/new-reg",
  "r9K_zwOyBJU": "https://community.letsencrypt.org/t/adding-random-entries-to-the-directory/33417",
  "revoke-cert": "https://acme-staging.api.letsencrypt.org/acme/revoke-cert"
}'
[Sat Feb 24 10:36:02 PST 2018] ACME_KEY_CHANGE='https://acme-staging.api.letsencrypt.org/acme/key-change'
[Sat Feb 24 10:36:02 PST 2018] ACME_NEW_AUTHZ='https://acme-staging.api.letsencrypt.org/acme/new-authz'
[Sat Feb 24 10:36:02 PST 2018] ACME_NEW_ORDER='https://acme-staging.api.letsencrypt.org/acme/new-cert'
[Sat Feb 24 10:36:02 PST 2018] ACME_NEW_ACCOUNT='https://acme-staging.api.letsencrypt.org/acme/new-reg'
[Sat Feb 24 10:36:02 PST 2018] ACME_REVOKE_CERT='https://acme-staging.api.letsencrypt.org/acme/revoke-cert'
[Sat Feb 24 10:36:02 PST 2018] ACME_AGREEMENT='https://letsencrypt.org/documents/LE-SA-v1.2-November-15-2017.pdf'
[Sat Feb 24 10:36:02 PST 2018] ACME_NEW_NONCE
[Sat Feb 24 10:36:02 PST 2018] ACME_VERSION
[Sat Feb 24 10:36:02 PST 2018] Le_NextRenewTime='1524550438'
[Sat Feb 24 10:36:02 PST 2018] OK
[Sat Feb 24 10:36:02 PST 2018] 1:Le_Domain='home.miscbitbag.org'
[Sat Feb 24 10:36:02 PST 2018] OK
[Sat Feb 24 10:36:02 PST 2018] 2:Le_Alt='no'
[Sat Feb 24 10:36:02 PST 2018] OK
[Sat Feb 24 10:36:02 PST 2018] 3:Le_Webroot='no'
[Sat Feb 24 10:36:02 PST 2018] OK
[Sat Feb 24 10:36:02 PST 2018] 4:Le_PreHook=''
[Sat Feb 24 10:36:03 PST 2018] OK
[Sat Feb 24 10:36:03 PST 2018] 5:Le_PostHook=''
[Sat Feb 24 10:36:03 PST 2018] OK
[Sat Feb 24 10:36:03 PST 2018] 6:Le_RenewHook=''
[Sat Feb 24 10:36:03 PST 2018] OK
[Sat Feb 24 10:36:03 PST 2018] 7:Le_LocalAddress='75.83.231.181'
[Sat Feb 24 10:36:03 PST 2018] OK
[Sat Feb 24 10:36:03 PST 2018] 8:Le_API='https://acme-staging.api.letsencrypt.org/directory'
[Sat Feb 24 10:36:03 PST 2018] _on_before_issue
[Sat Feb 24 10:36:03 PST 2018] 'no' contains 'no'
[Sat Feb 24 10:36:03 PST 2018] socat exists=0
[Sat Feb 24 10:36:03 PST 2018] Le_LocalAddress='75.83.231.181'
[Sat Feb 24 10:36:03 PST 2018] Check for domain='home.miscbitbag.org'
[Sat Feb 24 10:36:03 PST 2018] _currentRoot='no'
[Sat Feb 24 10:36:03 PST 2018] Standalone mode.
[Sat Feb 24 10:36:03 PST 2018] _checkport='80'
[Sat Feb 24 10:36:03 PST 2018] _checkaddr='75.83.231.181'
[Sat Feb 24 10:36:03 PST 2018] ss exists=1
[Sat Feb 24 10:36:03 PST 2018] netstat exists=0
[Sat Feb 24 10:36:03 PST 2018] Using: netstat
[Sat Feb 24 10:36:03 PST 2018] busybox netstat -ntl
[Sat Feb 24 10:36:03 PST 2018] netprc: checking any address
[Sat Feb 24 10:36:04 PST 2018] tcp        0      0 127.0.0.1:80            0.0.0.0:*               LISTEN
tcp        0      0 192.168.50.1:80         0.0.0.0:*               LISTEN
[Sat Feb 24 10:36:04 PST 2018] tcp port 80 is already used by
[Sat Feb 24 10:36:04 PST 2018] Please stop it first
[Sat Feb 24 10:36:04 PST 2018] _on_before_issue.

**Comment by FlyveHest** [comment]
I can attest to this, when _on_before is checked, it checks if anything is listening on for instance port 443, and not on the interface you specified with --local-address

Case in point, my box has 5 IPs (lets say 192.168.0.1-5), and I have a webserver running on 192.168.0.1 both port 80 and 443

When I run this command

`acme.sh --test --issue -d test.domain.tld --alpn --local-address 192.168.0.5`

I get the following on-before error

`LISTEN   0         128            192.168.0.1:443              0.0.0.0:*        users:(("apache2",pid=6043,fd=6),("apache2",pid=5410,fd=6),("apache2",pid=5408,fd=6),("apache2",pid=5406,fd=6),("apache2",pid=5405,fd=6),("apache2",pid=5404,fd=6),("apache2",pid=1678,fd=6))
tcp port 443 is already used by (("apache2",pid=6043,fd=6),("apache2",pid=5410,fd=6),("apache2",pid=5408,fd=6),("apache2",pid=5406,fd=6),("apache2",pid=5405,fd=6),("apache2",pid=5404,fd=6),("apache2",pid=1678,fd=6))
Please stop it first
_on_before_issue.`

As I specify the IP directly, it should check for that IP only.

**Comment by FlyveHest** [comment]
It looks like the ss command might return 0.0.0.0:* in the "Peer Address:Port" column, maybe this happens when you have multiple IPs on the same interface? (I've never used ss before)

Trying `ss -ntpl` on a system with only one IP on the interface gives me *:* in that column

**Comment by stevebovy** [comment]
Thanks for your recent comments.   I have not looked at this for a long time because I have already
"fixed" the problems by updating my private copy of the "script" :)  

But I would like to Point you to the following in the debug trace above:

[Sat Feb 24 10:36:03 PST 2018] ss exists=1
[Sat Feb 24 10:36:03 PST 2018] netstat exists=0
[Sat Feb 24 10:36:03 PST 2018] Using: netstat
[Sat Feb 24 10:36:03 PST 2018] busybox netstat -ntl
[Sat Feb 24 10:36:03 PST 2018] netprc: checking any address
[Sat Feb 24 10:36:04 PST 2018] tcp 0 0 127.0.0.1:80 0.0.0.0:* LISTEN
tcp 0 0 192.168.50.1:80 0.0.0.0:* LISTEN

I do not know what the (ss) command is but; within the context of the discussion of this issue it is
not relevant because it is not being used.  The command that is being used is "netstat"

Thanks again for your contribution to this discussion

**Comment by neilpang** [comment]
Closing as stale: no activity for over 18 months. If this still applies on the current release, please open a new issue with the `--debug 2` log.

**Comment by neilpang** [comment]
Fixed in 5cc1d952: the any-address fallback now greps "$LOCAL_ANY_ADDRESS:$_checkport", so the peer-address column no longer causes false "port already used" errors with --local-address.
